Rework documentation (#563)
cap-java/cds-feature-attachments: README.md

and handling by using an aspect Attachments.

It supports the [AWS, Azure and Google object stores](storage-targets/cds-feature-attachments-oss) and can connect to a [malware scanner](#malware-scanner).

## Table of Contents

<!-- TOC -->

* [Getting Started](#getting-started)
  * [Changes in the CDS Models](#changes-in-the-cds-models)
  * [UI](#ui)
  * [Storage Targets](#storage-targets)
  * [Malware Scanner](#malware-scanner)
  * [Outbox](#outbox)
  * [Restore Endpoint](#restore-endpoint)
    * [Motivation](#motivation)
    * [HTTP Endpoint](#http-endpoint)
    * [Security](#security)
* [Releases: Maven Central and Artifactory](#releases-maven-central-and-artifactory)
* [Minimum UI and CAP Java Version](#minimum-ui5-and-cap-java-version)
* [Architecture Overview](#architecture-overview)
  * [Design](#design)
  * [Multitenancy](#multitenancy)
  * [Object Stores](#object-stores)
  * [Model Texts](#model-texts)
* [Monitoring & Logging](#monitoring--logging)
* [Support, Feedback, Contributing](#support-feedback-contributing)
* [References & Links](#references--links)

> ## Getting Started

Reviewer: As someone who hasn't used the Java version of the plugin before, the length and amount of detail in this Getting Started section is a bit daunting. The actual setup steps are somewhat difficult to distinguish from the information as to how it works. Perhaps we should streamline it down to the absolute basics to get a simple example working and then explain more in-depth what is happening in a new section.

and handling by using an aspect Attachments.

It supports the [AWS, Azure and Google object stores](storage-targets/cds-feature-attachments-oss) and can connect to a [malware scanner](#malware-scanner).

## Table of Contents

<!-- TOC -->

* [Quick Start](#quick-start)
* [Usage](#usage)
  * [Changes in the CDS Models and for the UI](#changes-in-the-cds-models-and-for-the-UI)
  * [Storage Targets](#storage-targets)
  * [Malware Scanner](#malware-scanner)
  * [Outbox](#outbox)
  * [Restore Endpoint](#restore-endpoint)
    * [Motivation](#motivation)
    * [HTTP Endpoint](#http-endpoint)
    * [Security](#security)
* [Releases: Maven Central and Artifactory](#releases-maven-central-and-artifactory)
* [Minimum UI and CAP Java Version](#minimum-ui5-and-cap-java-version)
* [Architecture Overview](#architecture-overview)
  * [Design](#design)
  * [Multitenancy](#multitenancy)
  * [Object Stores](#object-stores)
  * [Model Texts](#model-texts)
* [Monitoring & Logging](#monitoring--logging)
* [Support, Feedback, Contributing](#support-feedback-contributing)
* [References & Links](#references--links)

> ## Quick Start

Reviewer: Quick start is much cleaner and easier to follow now!

For a quick setup with in-memory storage:

> - add the `cds-feature-attachments` Maven dependency to your `srv/pom.xml` and configure the `cds-maven-plugin` with the `resolve` goal as described in [Usage](#usage)

Reviewer: Capitalize 'Add' and 'Extend' at the start of the bullet points. Additionally, it's inconsistent in the README right now if we're ending bullet points with periods

- extend your CDS model with the `Attachments` aspect and annotate your service for UI integration as explained in [Changes in the CDS Models and for the UI](#changes-in-the-cds-models-and-for-the-UI)

> See the [incidents app](https://github.com/cap-java/incidents-app/) for a practical example.

Reviewer: Practical example of what? I would re-word this to clarify this is sample illustrates the plugin very well

For object store integration, see [Amazon, Azure, and Google Object Stores](storage-targets/cds-feature-attachments-oss).

## Usage

> 

Reviewer: Do we want a subsection here for maven setup? Currently we reference this section using 'Usage'

</dependency>
```

> To be able to also use the CDS models defined in this plugin the `cds-maven-plugin` needs to be used with the `resolve` goal to make the CDS models available in the project:

Reviewer: "In order to use the CDS models..."

Depending where the aspect `Attachments` shall be used in the application's CDS model, different approaches need to be implemented.
- If the aspect `Attachments` shall be used on an entity provided in the `db` module, the corresponding entity needs to be extended from a CDS file in the `srv` module. Therefore the entity from the `db` folder needs to be imported with an `using` statement. Then, this entity can be extended with a new field that is a `Composition of many Attachments`.
 The following example shows how to extend the `db` entity `Books` in a CDS file in the `srv` module:

  ```cds
  using {my.bookshop as my} from '../db/index';
  using {sap.attachments.Attachments} from 'com.sap.cds/cds-feature-attachments';

  // Extends the Books entity with the Attachments composition
  extend my.Books with {
    covers : Composition of many Attachments;
  };
  ```

- To use the aspect `Attachments` in the `srv` module, the following code needs to be added to the existing entity definition:

  ```cds
  using {sap.attachments.Attachments} from `com.sap.cds/cds-feature-attachments`;

  entity Items : cuid {
      ...
      attachments : Composition of many Attachments;
      ...
  }

>   ```

Reviewer 2: I also struggled here. "existing entity definition" sounds like _db/schema.cds_. 
I also struggled to understand the relation of the entity Items with the previous snippet where Books are extended. On a second (third 😅) read I understood that those are separate scenarios and that there is no relation.
Should we only show the "entity in db module" scenario?

      ...,
      {
        $Type  : 'UI.ReferenceFacet',
        ID     : 'AttachmentsFacet',
        Label  : '{i18n>attachments}',
        Target : 'attachments/@UI.LineItem'
      }
    ]
  );
  ```

The UI Facet can also be added directly after other UI Facets in a `cds` file in the `app` folder.

### Storage Targets

By default, the plugin operates without a dedicated storage target, storing attachments directly in the [underlying database](cds-feature-attachments/src/main/resources/cds/com.sap.cds/cds-feature-attachments/attachments.cds#L17).

Other available storage targets:
- [Amazon, Azure and Google Object Stores](storage-targets/cds-feature-attachments-oss)
- [local file system as a storage backend](storage-targets/cds-feature-attachments-fs) (only for testing scenarios)

When using a dedicated storage target, the attachment is not stored in the underlying database; instead, it is saved on the specified storage target, and only a reference to the file is kept in the database, as defined in the [CDS model](cds-feature-attachments/src/main/resources/cds/com.sap.cds/cds-feature-attachments/attachments.cds#L20).

### Malware Scanner

This plugin checks for a binding to

> the [SAP Malware Scanning Service](https://help.sap.com/docs/malware-scanning-servce), this needs to have the label `malware-scanner`. The entry in the [mta-file](https://cap.cloud.sap/docs/guides/deployment/to-cf#add-mta-yaml) possibly looks like:

Reviewer: "The entry in the mta-file may look like:

Scan status codes:
- `Clean`: Only attachments with the status `Clean` are accessible.
- `Scanning`: Immediately after upload, the attachment is marked as `Scanning`. Depending on processing speed, it may already appear as `Clean` when the page is reloaded.
- `Unscanned`: Attachment is still unscanned.
- `Failed`: Scanning failed.
- `Infected`: The attachment is infected.

### Outbox

In this plugin the [persistent outbox](https://cap.cloud.sap/docs/java/outbox#persistent) is used to mark attachments as
deleted.
When using this plugin, the persistent outbox is enabled by default.
In the capire documentation of the [persistent outbox](https://cap.cloud.sap/docs/java/outbox#persistent) is it
described how to overwrite
the default outbox configuration.

If the default shall be used, nothing needs to be done.

### Restore Endpoint

The attachment service has an event `RESTORE_ATTACHMENTS`.
This event can be called with a timestamp to restore externally stored attachments.

#### Motivation

Documents which are marked as deleted can be restored.

The use cases behind this feature are:
- Restoring attachments after a database backup is restored:

> When you restore a database backup, any attachments stored in external storage (object stores, etc.) also need to be restored to maintain data consistency.

Reviewer: This is not appearing in the README as a new line. Plus let's remove "you" with : "When a database backup is restored, ..."

      - name: cf-logging
      - name: malware-scanner
...
resources:
  ...
  - name: malware-scanner
    type: org.cloudfoundry.managed-service
    parameters:
      service: malware-scanner
      service-plan: clamav
```

The malware scanner is used in the `AttachmentService` to scan
attachments.
If the default implementation of this service is overwritten, e.g. by using the plugin enhancement of the
[SAP Document Management Service](https://help.sap.com/docs/document-management-service), then the overwriting plugin
is responsible for the malware scan.

If there is no malware scanner available the attachments are automatically marked as `Clean`.

Scan status codes:
- `Clean`: Only attachments with the status `Clean` are accessible.
- `Scanning`: Immediately after upload, the attachment is marked as `Scanning`. Depending on processing speed, it may already appear as `Clean` when the page is reloaded.
- `Unscanned`: Attachment is still unscanned.
- `Failed`: Scanning failed.
- `Infected`: The attachment is infected.

### Outbox

In this plugin the [persistent outbox](https://cap.cloud.sap/docs/java/outbox#persistent) is used to mark attachments as
deleted.

> The enablement of the outbox is also included in the cds models of this plugin.

Reviewer 2: ```suggestion
When using this plugin, the persistent outbox is enabled by default.
```

In the capire documentation of the [persistent outbox](https://cap.cloud.sap/docs/java/outbox#persistent) is it
described how to overwrite
the default outbox configuration.

If the default shall be used, nothing needs to be done.

### Restore Endpoint

The attachment service has an event `RESTORE_ATTACHMENTS`.

> This event can be called with a timestamp to restore externally stored attachments.

Reviewer 2: This section shows how to create a service (L260) and only later it is secured (L316). Can we merge this into the initial service definition and already there mention that it is / should be secured?

#### Motivation

> 

Reviewer: Are the use cases a list of reasons? Would be clearer to turn this into bullet points if so

#### Security

> To secure the endpoint security annotations can be used, for example:

Reviewer: "To secure the endpoint, security annotations can be used. For example:" (or should it be "To secure the endpoint security, annotations can be used."?)

> 

Reviewer: Above, the meaning is vague. Are they saying there are multiple options for annotations or that multiple annotations can be used at the same time? A better sentence would be "Various other annotations can be used to secure the service" (multiple annotation options) or "Other annotations can be used in addition to secure the service" (multiple can be used at the same time)

More information about the CAP Java security concept can be found in
the [CAP Java Documentation](https://cap.cloud.sap/docs/java/security).

## Releases: Maven Central and Artifactory

- The plugin is released to Maven Central at: https://central.sonatype.com/artifact/com.sap.cds/cds-feature-attachments (public access)

- See the [changelog](./doc/CHANGELOG.md) for the latest changes.

> - If you want to test snapshot versions of this plugin, you need to configure the Artifactory in your `${HOME}/.m2/settings.xml`. See [the maven settings](https://maven.apache.org/settings.html#Repositories) for further details.

Reviewer: Remove "you" here

More information about the CAP Java security concept can be found in
the [CAP Java Documentation](https://cap.cloud.sap/docs/java/security).

## Releases: Maven Central and Artifactory

- The plugin is released to Maven Central at: https://central.sonatype.com/artifact/com.sap.cds/cds-feature-attachments

> - Snapshots are deployed to SAP's Artifactory in DMZ: https://common.repositories.cloud.sap/artifactory/cap-java/com/sap/cds/cds-feature-attachments/

Reviewer 2: Are those available externally?

Author: https://central.sonatype.com/artifact/com.sap.cds/cds-feature-attachments is, https://common.repositories.cloud.sap/artifactory/cap-java/com/sap/cds/cds-feature-attachments/  is not, I have added that info.
